### PR TITLE
Customizable Banners!

### PR DESCRIPTION
**Pull request by MatheusIsNotAvailable**
[FEATURE] [WIKI]

## About the PR 
Adds Banners which can be made by using cloth on a wall, the banner inherits the cloth's material, the banner can then be colored by using a crayon on it, after it has been colored you can draw various patterns or letters on it, there is no limit to how many you can draw they are the same color as the crayon you used to draw them too, you can clear it of it's color and drawings by click dragging it, and you can remove it completely by using a wire cutter or similar on it which will give you the cloth you used back. Also makes a small change to rainbow crayons so they work as you expect them to when used in a banner.
## Why's this needed? 
It is cool, lets players customize places in a cool way
Banners!

## Changelog

```changelog
(u)Arthur Holiday
(*)Added Customizable banners, use cloth on a wall to make one, color and draw on it with crayons.
```

### CODE REVIEW

**Comment by Gannets** [comment]
Can't speak to the code quality but this is a really nice idea, sprites look good too. 👍

File: code/obj/decal/poster.dm [hunk 1]
```diff
@@ -979,3 +979,64 @@
...
 		C.Browse("<img src=\"[resource("images/pw_map.png")]\">","window=Map;size=[imgw]x[imgh];title=Map")
+
+/obj/decal/poster/banner
+	name = "\proper banner"
+	desc = "An unfinished banner, try adding some color to it by using a crayon!"
+	icon = 'icons/obj/decals/banners.dmi'
+	icon_state = "banner_base"
+	popup_win = 0
+	flags = TGUI_INTERACTIVE
+	var/colored = 0
```
Comment (PrimeNumb) [inline code/obj/decal/poster.dm:990]:
```suggestion
	var/colored = FALSE
```

File: code/obj/decal/poster.dm [hunk 2]
```diff
@@ -979,3 +979,64 @@
...
+	var/chosen_overlay
+	var/static/list/choosable_overlays = list("Horizontal Stripes","Vertical Stripes","Diagonal Stripes","Cross","Diagonal Cross","Full","Full Gradient",
+	"Left Line","Middle Line","Right Line","Northwest Line","Northeast Line","Southwest Line","Southeast Line","Big Ball","Medium Ball","Small Ball",
+	"A","B","C","D","E","F","G","H","I","J","K","L","M","N","O","P","Q","R","S","T","U","V","W","X","Y","Z")
+
+	proc/clear()
+		if(src.material) src.color = src.material.color
+		else src.color = "#ffffff" // In case the material is null
+		src.overlays = null
+		src.colored = 0
```
Comment (PrimeNumb) [inline code/obj/decal/poster.dm:1000]:
```suggestion
		src.colored = FALSE
```

File: code/obj/decal/poster.dm [hunk 3]
```diff
@@ -979,3 +979,64 @@
...
+	New()
+		..()
+		var/static/image/banner_holder = image(src.icon, , "banner_holder")
+		banner_holder.appearance_flags = RESET_COLOR
+		src.underlays.Add(banner_holder)
+		return
+
+	attackby(obj/item/W, mob/user)
+		if(istype(W,/obj/item/pen/crayon))
+			if(src.colored == 1)
```
Comment (PrimeNumb) [inline code/obj/decal/poster.dm:1012]:
```suggestion
			if(src.colored == TRUE)
```

File: code/obj/decal/poster.dm [hunk 4]
```diff
@@ -979,3 +979,64 @@
...
+				var/mutable_appearance/new_overlay = mutable_appearance(src.icon, chosen_overlay)
+				new_overlay.appearance_flags = RESET_COLOR
+				new_overlay.color = W.color
+				src.overlays.Add(new_overlay)
+				logTheThing("station", user, null, "Drew a [chosen_overlay] in [src] with [W] at [log_loc(user)].")
+				desc = "A banner, colored and decorated"
+
+			if(src.colored == 0)
+				src.color = W.color
+				src.colored = 1
```
Comment (PrimeNumb) [inline code/obj/decal/poster.dm:1024]:
```suggestion
				src.colored = TRUE
```

**Review by PrimeNumb — CHANGES_REQUESTED** [review]

File: code/obj/decal/poster.dm [hunk 5]
```diff
@@ -979,3 +979,64 @@
 			return
 
 		C.Browse("<img src=\"[resource("images/pw_map.png")]\">","window=Map;size=[imgw]x[imgh];title=Map")
+
+/obj/decal/poster/banner
+	name = "\proper banner"
+	desc = "An unfinished banner, try adding some color to it by using a crayon!"
+	icon = 'icons/obj/decals/banners.dmi'
+	icon_state = "banner_base"
+	popup_win = 0
```
Comment (PrimeNumb) [inline code/obj/decal/poster.dm:988]: looks like `popup_win` is never used?

File: code/obj/decal/poster.dm [hunk 6]
```diff
@@ -979,3 +979,64 @@
...
+				chosen_overlay = tgui_input_list(user, "What do you want to draw?", null, choosable_overlays)
+				if (!chosen_overlay) return
+				var/mutable_appearance/new_overlay = mutable_appearance(src.icon, chosen_overlay)
+				new_overlay.appearance_flags = RESET_COLOR
+				new_overlay.color = W.color
+				src.overlays.Add(new_overlay)
+				logTheThing("station", user, null, "Drew a [chosen_overlay] in [src] with [W] at [log_loc(user)].")
+				desc = "A banner, colored and decorated"
+
+			if(src.colored == 0)
```
Comment (PrimeNumb) [inline code/obj/decal/poster.dm:1022]:
```suggestion
			if(src.colored == FALSE)
```

**Review by PrimeNumb — CHANGES_REQUESTED** [review]

File: code/modules/materials/Mat_RawMaterials.dm [hunk 7]
```diff
@@ -177,6 +177,16 @@
 			else
 				return ..()
 
+		afterattack(turf/simulated/A, mob/user)
+			if(locate(/obj/decal/poster/banner,A))
```
Comment (ZeWaka) [inline code/modules/materials/Mat_RawMaterials.dm:181]:
```suggestion
			if(locate(/obj/decal/poster/banner, A))
```
consteienycy

File: code/modules/materials/Mat_RawMaterials.dm [hunk 8]
```diff
@@ -177,6 +177,16 @@
...
+				return
+			else if(istype(A, /turf/simulated/wall/))
+				var/obj/decal/poster/banner/B = new(A)
+				if(src.material) B.setMaterial(src.material)
+				logTheThing("station", user, null, "Made [B] at [log_loc(user)].")
```
Comment (ZeWaka) [inline code/modules/materials/Mat_RawMaterials.dm:186]: weirdly worded please be more clear

File: code/modules/materials/Mat_RawMaterials.dm [hunk 9]
```diff
@@ -177,6 +177,16 @@
...
+				src.change_stack_amount(-1)
+				user.visible_message("<span class='notice'>[user] Makes [B]!.</span>")
```
Comment (ZeWaka) [inline code/modules/materials/Mat_RawMaterials.dm:188]: poorly worded

File: code/obj/decal/poster.dm [hunk 10]
```diff
@@ -979,3 +979,63 @@
 			return
 
 		C.Browse("<img src=\"[resource("images/pw_map.png")]\">","window=Map;size=[imgw]x[imgh];title=Map")
+
+/obj/decal/poster/banner
+	name = "\proper banner"
```
Comment (ZeWaka) [inline code/obj/decal/poster.dm:984]: if it's proper why do you have it lowercase

File: code/obj/decal/poster.dm [hunk 11]
```diff
@@ -979,3 +979,63 @@
...
+	desc = "An unfinished banner, try adding some color to it by using a crayon!"
+	icon = 'icons/obj/decals/banners.dmi'
+	icon_state = "banner_base"
+	popup_win = 0
+	flags = TGUI_INTERACTIVE
```
Comment (ZeWaka) [inline code/obj/decal/poster.dm:989]: doubtful

File: code/obj/item/pens_writing_etc.dm [hunk 12]
```diff
@@ -379,14 +379,22 @@
...
 		write_on_turf(var/turf/T as turf, var/mob/user as mob, params)
 			if (!T || !user || src.in_use || get_dist(T, user) > 1)
 				return
 			src.font_color = random_saturated_hex_color(1)
 			src.color_name = hex2color_name(src.font_color)
+			src.color = src.font_color
 			..()
 
+		afterattack(obj/target, mob/user)
+			if (istype(target, /obj/decal/poster/banner)) // so they alter color when used on a banner
```
Comment (ZeWaka) [inline code/obj/item/pens_writing_etc.dm:393]: why are you doing afterattack and not attackby on the actual banner object - this block of code does not make sense to me

File: code/obj/decal/poster.dm [hunk 13]
```diff
@@ -979,3 +979,63 @@
...
+	var/colored = FALSE
+	var/chosen_overlay
+	var/static/list/choosable_overlays = list("Horizontal Stripes","Vertical Stripes","Diagonal Stripes","Cross","Diagonal Cross","Full","Full Gradient",
+	"Left Line","Middle Line","Right Line","Northwest Line","Northeast Line","Southwest Line","Southeast Line","Big Ball","Medium Ball","Small Ball",
+	"A","B","C","D","E","F","G","H","I","J","K","L","M","N","O","P","Q","R","S","T","U","V","W","X","Y","Z","0","1","2","3","4","5","6","7","8","9","+","-","=")
+
+	proc/clear()
```
Comment (ZeWaka) [inline code/obj/decal/poster.dm:996]: clear what?

File: code/obj/decal/poster.dm [hunk 14]
```diff
@@ -979,3 +979,63 @@
...
+		if(src.material) src.color = src.material.color
+		else src.color = "#ffffff" // In case the material is null
```
Comment (ZeWaka) [inline code/obj/decal/poster.dm:998]: indent your blocks

File: code/obj/decal/poster.dm [hunk 15]
```diff
@@ -979,3 +979,63 @@
...
+		src.overlays = null
+		src.colored = FALSE
+		usr.visible_message("<span class='notice'>[usr] Clears the [src].</span>")
+
+	New()
+		..()
+		var/static/image/banner_holder = image(src.icon, , "banner_holder")
+		banner_holder.appearance_flags = RESET_COLOR
+		src.underlays.Add(banner_holder)
+		return
```
Comment (ZeWaka) [inline code/obj/decal/poster.dm:1008]:
```suggestion
```
uselsss

File: code/obj/decal/poster.dm [hunk 16]
```diff
@@ -979,3 +979,63 @@
...
+	proc/clear()
+		if(src.material) src.color = src.material.color
+		else src.color = "#ffffff" // In case the material is null
+		src.overlays = null
+		src.colored = FALSE
+		usr.visible_message("<span class='notice'>[usr] Clears the [src].</span>")
+
+	New()
+		..()
+		var/static/image/banner_holder = image(src.icon, , "banner_holder")
```
Comment (ZeWaka) [inline code/obj/decal/poster.dm:1005]: umm why is there a blank argument what's going on here!!!

File: code/obj/decal/poster.dm [hunk 17]
```diff
@@ -979,3 +979,63 @@
...
+
+	proc/clear()
+		if(src.material) src.color = src.material.color
+		else src.color = "#ffffff" // In case the material is null
+		src.overlays = null
+		src.colored = FALSE
+		usr.visible_message("<span class='notice'>[usr] Clears the [src].</span>")
+
+	New()
+		..()
```
Comment (ZeWaka) [inline code/obj/decal/poster.dm:1004]:
```suggestion
		. = ..()
```

File: code/obj/decal/poster.dm [hunk 18]
```diff
@@ -979,3 +979,63 @@
...
+		var/static/image/banner_holder = image(src.icon, , "banner_holder")
+		banner_holder.appearance_flags = RESET_COLOR
+		src.underlays.Add(banner_holder)
+		return
+
+	attackby(obj/item/W, mob/user)
+		if(istype(W,/obj/item/pen/crayon))
+			if(src.colored == TRUE)
```
Comment (ZeWaka) [inline code/obj/decal/poster.dm:1012]:
```suggestion
			if(src.colored)
```
uncnessary

File: code/obj/decal/poster.dm [hunk 19]
```diff
@@ -979,3 +979,63 @@
...
+				chosen_overlay = tgui_input_list(user, "What do you want to draw?", null, choosable_overlays)
```
Comment (ZeWaka) [inline code/obj/decal/poster.dm:1013]: null? i like titles on my windows don't you?

File: code/obj/decal/poster.dm [hunk 20]
```diff
@@ -979,3 +979,63 @@
...
+				if (!chosen_overlay) return
+				var/mutable_appearance/new_overlay = mutable_appearance(src.icon, chosen_overlay)
+				new_overlay.appearance_flags = RESET_COLOR
+				new_overlay.color = W.color
+				src.overlays.Add(new_overlay)
+				logTheThing("station", user, null, "Drew a [chosen_overlay] in [src] with [W] at [log_loc(user)].")
+				desc = "A banner, colored and decorated"
+
+			if(src.colored == FALSE)
```
Comment (ZeWaka) [inline code/obj/decal/poster.dm:1022]:
```suggestion
			else
```

File: code/obj/decal/poster.dm [hunk 21]
```diff
@@ -979,3 +979,63 @@
...
+			qdel(src)
+
+	MouseDrop(atom/over_object, src_location, over_location)
+		..()
+		if (usr.stat || usr.restrained() || !can_reach(usr, src))
+			return
+
+		else if(tgui_alert(usr, "Are you sure you want to clear the banner?","Confirmation",list("Yes","No")) == "Yes")
+			clear()
+		else return
```
Comment (ZeWaka) [inline code/obj/decal/poster.dm:1041]: please indent your blocks after a control statement

File: code/obj/decal/poster.dm [hunk 22]
```diff
@@ -979,3 +979,63 @@
...
+			if(src.material) C.setMaterial(src.material)
+			else C.setMaterial(getMaterial("cotton")) // In case the material is null
+			qdel(src)
+
+	MouseDrop(atom/over_object, src_location, over_location)
+		..()
+		if (usr.stat || usr.restrained() || !can_reach(usr, src))
+			return
+
+		else if(tgui_alert(usr, "Are you sure you want to clear the banner?","Confirmation",list("Yes","No")) == "Yes")
```
Comment (ZeWaka) [inline code/obj/decal/poster.dm:1039]: please be consistent in your usage of spaces for arguments
```suggestion
		else if(tgui_alert(usr, "Are you sure you want to clear the banner?", "Confirmation", list("Yes", "No")) == "Yes")
```

File: code/obj/decal/poster.dm [hunk 23]
```diff
@@ -979,3 +979,63 @@
...
+				src.overlays.Add(new_overlay)
+				logTheThing("station", user, null, "Drew a [chosen_overlay] in [src] with [W] at [log_loc(user)].")
+				desc = "A banner, colored and decorated"
+
+			if(src.colored == FALSE)
+				src.color = W.color
+				src.colored = TRUE
+				desc = "A colored banner, try adding some drawings to it with a crayon!"
+
+		if(istool(W,TOOL_SNIPPING| TOOL_CUTTING | TOOL_SAWING))
```
Comment (ZeWaka) [inline code/obj/decal/poster.dm:1027]:
```suggestion
		if(istool(W,TOOL_SNIPPING | TOOL_CUTTING | TOOL_SAWING))
```
AUGH

File: code/obj/decal/poster.dm [hunk 24]
```diff
@@ -979,3 +979,63 @@
...
+			user.visible_message("<span class='notice'>[user] cuts off [src] with [W].</span>")
+			var/obj/item/material_piece/cloth/C = new(user.loc)
+			if(src.material) C.setMaterial(src.material)
+			else C.setMaterial(getMaterial("cotton")) // In case the material is null
```
Comment (ZeWaka) [inline code/obj/decal/poster.dm:1031]: spaces after multipronged control statements plase

File: code/obj/decal/poster.dm [hunk 25]
```diff
@@ -979,3 +979,63 @@
...
+
+	attackby(obj/item/W, mob/user)
+		if(istype(W,/obj/item/pen/crayon))
+			if(src.colored == TRUE)
+				chosen_overlay = tgui_input_list(user, "What do you want to draw?", null, choosable_overlays)
+				if (!chosen_overlay) return
+				var/mutable_appearance/new_overlay = mutable_appearance(src.icon, chosen_overlay)
+				new_overlay.appearance_flags = RESET_COLOR
+				new_overlay.color = W.color
+				src.overlays.Add(new_overlay)
```
Comment (ZeWaka) [inline code/obj/decal/poster.dm:1018]: consider using overlay manager

**Review by ZeWaka — CHANGES_REQUESTED** [review]

**Review by Carbadox — COMMENTED** [review]
Tested locally and it works as intended.

My own suugestions would be to make that splashing a banner with acetone would reset it back to the original clean state and maybe also make acid destroy it. 

All in all though, very cool.

**Comment by MatheusIsNotAvailable** [comment]
Github mobile hid a bunch of the earlier reviews so I missed some stuff, gonna do the rest today probably, sorry for requesting a review when it isn't done yet

**Review by virvatuli-gh — COMMENTED** [review]
Looks good 2 me but I'll wait for Zewaka to check in before merge

**Review by ZeWaka — APPROVED** [review]

**Pull request merged**